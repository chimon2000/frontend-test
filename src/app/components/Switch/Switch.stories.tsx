import * as React from "react";
import { storiesOf } from "@storybook/react";
import { withInfo } from "@storybook/addon-info";
import { action } from "@storybook/addon-actions";
import { boolean, text } from "@storybook/addon-knobs";

import { Switch } from "./Switch";

const stories = storiesOf("Components", module);

stories.add(
  "Switch",
  withInfo({ inline: true })(() => (
    <Switch
      onChange={action("Change event")}
      checked={boolean("Checked", true)}
      disabled={boolean("Disabled", false)}
      name="switch_test"
      id="switch_test"
    />
  ))
);
