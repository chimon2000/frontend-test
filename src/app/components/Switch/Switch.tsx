import * as React from "react";

import { joinClassList } from "../../helpers";

export interface ChangeEventObject {
  name: string;
  value: boolean;
}
export interface SwitchProps {
  /**
   * Triggers when Checkbox is toggled checked/unchecked.
   */
  onChange: (result: ChangeEventObject) => void;
  /**
   * The Name of the checkbox. Make sure it's unique to the page!
   */
  name: string;
  checked: boolean;
  disabled?: boolean;
  /**
   * Can be text or a React Element.
   */
  label?: [string, string];
  children?: React.ReactNode;
  /** Required for inputs. */
  id: string;
  className?: string;
}

export const Switch: React.FunctionComponent<SwitchProps> = ({
  onChange,
  name,
  checked,
  disabled = false,
  label = ["", ""],
  children,
  id,
  className
}) => {
  const [isChecked, toggleChecked] = React.useState(checked);
  const classes = [
    `switch`,
    `${className || ""}`,
    `${checked ? "switch--on" : ""}`
  ];

  React.useEffect(() => toggleChecked(checked), [checked]);

  return (
    <div className={joinClassList(classes)}>
      <label className="switch__label" htmlFor={id}>
        {label[0]}
        <input
          className="switch__input"
          type="checkbox"
          name={name}
          id={id}
          checked={isChecked}
          disabled={disabled}
          onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
            toggleChecked(!isChecked);
            onChange({ name, value: e.target.checked });
          }}
        />
        <span className="switch__lever" />
        {label[1]}
      </label>
    </div>
  );
};
