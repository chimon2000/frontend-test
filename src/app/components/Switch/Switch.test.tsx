import * as React from "react";
import { shallow } from "enzyme";

import { Switch } from "./Switch";

describe("<Switch />", () => {
  it("doesn't blow up", () => {
    const wrapper = shallow(
      <Switch
        onChange={() => {}}
        checked={true}
        disabled={false}
        name="switch_test"
        id="switch_test"
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("Triggers onChange when change happens", () => {
    let triggered = false;
    const updateMethod = () => {
      triggered = true;
    };
    const wrapper = shallow(
      <Switch
        onChange={updateMethod}
        checked={false}
        disabled={false}
        name="switch_test"
        id="switch_test"
      />
    );

    const input = wrapper.find(".switch__input");
    input.simulate("change", { target: { checked: true } });
    expect(triggered).toEqual(true);
  });

  it("Is checked if value is passed", () => {
    const wrapper = shallow(
      <Switch
        onChange={() => {}}
        checked={true}
        disabled={false}
        name="switch_test"
        id="switch_test"
      />
    );
    const input = wrapper.find(".switch__input").get(0);
    expect(input.props.checked).toBe(true);
  });
});
