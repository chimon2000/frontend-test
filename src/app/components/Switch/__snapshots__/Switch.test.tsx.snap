// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> doesn't blow up 1`] = `
<div
  className="switch switch--on"
>
  <label
    className="switch__label"
    htmlFor="switch_test"
  >
    <input
      checked={true}
      className="switch__input"
      disabled={false}
      id="switch_test"
      name="switch_test"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="switch__lever"
    />
  </label>
</div>
`;
